#pragma once
#include <QDialog>
#include "schema.h"

namespace Ui { class SchemaEditor; }

/**
 * @class SchemaEditor
 * @brief Діалогове вікно для створення та редагування однієї схеми.
 *
 * @details Клас надає користувачеві інтерфейс для визначення нової структури нотаток.
 * Він динамічно відображає список полів, які користувач додає, і служить шаблоном
 * для побудови об'єкта @c Schema.
 */
class SchemaEditor : public QDialog {
    Q_OBJECT

public:
    /**
     * @brief Конструктор для створення нової, порожньої схеми.
     * @details Ініціалізує діалог для створення абсолютно нової схеми; список полів
     * та поле назви залишаються порожніми.
     * @param parent Вказівник на батьківський віджет.
     */
    explicit SchemaEditor(QWidget *parent = nullptr);

    /**
     * @brief Конструктор для редагування існуючої схеми.
     * @details Ініціалізує діалог, завантажуючи існуючі назву та поля з об'єкта @c schemaToEdit
     * у тимчасове сховище для редагування.
     * @param schemaToEdit Схема, дані якої потрібно завантажити в редактор.
     * @param parent Вказівник на батьківський віджет.
     */
    explicit SchemaEditor(const Schema& schemaToEdit, QWidget *parent = nullptr);

    /**
     * @brief Деструктор класу.
     */
    ~SchemaEditor();

    /**
     * @brief Збирає дані з полів вводу та повертає готовий об'єкт Schema.
     * @details Викликається після успішної валідації і натискання кнопки "ОК".
     * Складає назву схеми з поля вводу та полів зі списку @c m_tempFields.
     * @return Створений або оновлений об'єкт Schema.
     */
    Schema getSchema() const;

    private slots:
    /**
     * @brief Слот, що викликається при натисканні на кнопку "Додати поле".
     *
     * @details Додає нове поле до @c m_tempFields і відображає його назву у списку.
     * Це є проміжним кроком перед фінальним створенням об'єкта @c Schema.
     */
    void on_addFieldButton_clicked();

    /**
     * @brief Перевизначений слот, що викликається при натисканні "OK".
     *
     * @details Виконує валідацію (перевірку), що назва схеми не порожня. Якщо
     * валідація успішна, діалог закривається.
     */
    void accept() override;

protected:
    /**
     * @brief Обробник подій для динамічного оновлення інтерфейсу.
     *
     * @details Перехоплює подію зміни мови (@c QEvent::LanguageChange)
     * для оновлення заголовка вікна та всіх елементів, що підтримують
     * інтернаціоналізацію (i18n).
     * @param event Подія, що надійшла від системи.
     */
    void changeEvent(QEvent *event) override;

private:
    Ui::SchemaEditor *ui;        ///< Вказівник на об'єкти інтерфейсу, згенеровані @c uic.

    /**
     * @brief Тимчасове сховище для полів, які додає користувач.
     * @details Поля зберігаються тут до фінального натискання "ОК",
     * після чого вони переносяться в об'єкт @c Schema.
     */
    QList<Field> m_tempFields;
};